File: mac/PabloCompanion/Models/SubscriptionStatus.swift
import Foundation

/// Wrapper for the `GET /api/users/me/status` response.
/// The `subscription` field is nil when the backend is not running in SaaS mode.
struct SubscriptionResponse: Codable, Sendable {
    let subscription: SubscriptionInfo?
}

/// Subscription details returned by the backend.
struct SubscriptionInfo: Codable, Sendable, Equatable {
    let status: SubscriptionState
    let plan: String
    let trialSessionsUsed: Int?
    let trialSessionsLimit: Int?
    let trialDaysLimit: Int?
    let trialStart: String?
    let graceExtensionAvailable: Bool
    let graceExtensionExpiresAt: String?
}

/// Subscription lifecycle states driven by the backend.
enum SubscriptionState: String, Codable, Sendable {
    case active
    case trial
    case pastDue = "past_due"
    case canceled
}

/// What the banner view should render. Computed from `SubscriptionInfo`.
enum SubscriptionBannerState: Equatable {
    case hidden
    case trial(sessionsRemaining: Int?, daysRemaining: Int?)
    case pastDue(extensionAvailable: Bool)
    case canceled(extensionAvailable: Bool)
    case graceActive(expiresAt: Date)
}

File: windows/PabloCompanion/Models/SubscriptionStatus.cs
using System.Text.Json.Serialization;

namespace PabloCompanion.Models;

/// <summary>
/// Wrapper for the GET /api/users/me/status response.
/// </summary>
public sealed record SubscriptionResponse(
    [property: JsonPropertyName("subscription")]
    SubscriptionInfo? Subscription);

/// <summary>
/// Subscription details returned by the backend.
/// </summary>
public sealed record SubscriptionInfo(
    [property: JsonPropertyName("status")]
    string Status,
    [property: JsonPropertyName("plan")]
    string Plan,
    [property: JsonPropertyName("trial_sessions_used")]
    int? TrialSessionsUsed,
    [property: JsonPropertyName("trial_sessions_limit")]
    int? TrialSessionsLimit,
    [property: JsonPropertyName("trial_days_limit")]
    int? TrialDaysLimit,
    [property: JsonPropertyName("trial_start")]
    string? TrialStart,
    [property: JsonPropertyName("grace_extension_available")]
    bool GraceExtensionAvailable,
    [property: JsonPropertyName("grace_extension_expires_at")]
    string? GraceExtensionExpiresAt);
